ntoskrnl/ke/bug.c

```diff
@@ -650,84 +774,147 @@ KiDisplayBlueScreen(IN ULONG MessageId,
-        InbvSolidColorFill(0, 0, 639, 479, 4);
-        InbvSetTextColor(15);
-        InbvInstallDisplayStringFilter(NULL);
-        InbvEnableDisplayString(TRUE);
-        InbvSetScrollRegion(0, 0, 639, 479);
+        /* Set up display for blur screen */
+        SetUpBlueScreen(BackgroundColor);
+        
+        /* TODO: Change background color. We can change the color by creating gradient in palette from X color to "White" color. */
     }
-
+    
+    /* Print out initial message */
+    KeGetBugMessageText(BUGCHECK_MESSAGE_INTRO, NULL);
+    InbvDisplayString("\r\n");
+    
     /* Check if this is a hard error */
     if (IsHardError)
     {
         /* Display caption and message */
-        if (HardErrCaption) InbvDisplayString(HardErrCaption);
-        if (HardErrMessage) InbvDisplayString(HardErrMessage);
+        if (HardErrCaption)
+        {
+            InbvDisplayString(HardErrCaption);
+            InbvDisplayString("\r\n");
+        }
+        if (HardErrMessage) 
+        {
+            InbvDisplayString(HardErrMessage);
+            InbvDisplayString("\r\n");
+        }
     }
-
-    /* Begin the display */
+    
     InbvDisplayString("\r\n");
+    
+    /* Check if this is the generic message and it's friendly bsod style */
+    if (MessageId == BUGCODE_PSS_MESSAGE && ErrorScreenStyle < 2)
+    {
+        /* It is, so get the bug code string as well */
+        KeGetBugMessageText((ULONG)KiBugCheckData[0], NULL);
+        InbvDisplayString("\r\n\r\n");
+    }
 
-    /* Print out initial message */
-    KeGetBugMessageText(BUGCHECK_MESSAGE_INTRO, NULL);
-    InbvDisplayString("\r\n\r\n");
-
+    /* Get the bug code string */
+    KeGetBugMessageText(MessageId, NULL);
+    InbvDisplayString("\r\n\r\n\r\n");
+    
+    
+    /* After main messages make the text's color less distracting. */
+    InbvSetTextColor(7);
+    
+    sprintf(AnsiName, "%s\r\n", NtBuildLab);
+    AppendAdditionalData(AnsiName, TRUE);
+    
+    /* ThFabba's code, why not to add this information? */
+    AppendAdditionalData("\r\n\r\n---stack backtrace---\r\n", DisplayBacktrace);
+    {
+        CHAR Buffer[64];
+        PVOID *Frame;
+        PVOID *FramePc, *NextFrame;
+        PLDR_DATA_TABLE_ENTRY LdrEntry;
+        BOOLEAN InSystem;
+        CHAR AnsiName[64];
+ 
+        NextFrame = _AddressOfReturnAddress();
+        NextFrame--;
+        do
+        {
+            Frame = NextFrame;
+            FramePc = Frame[1];
+            NextFrame = Frame[0];
+ 
+            if ((ULONG_PTR)FramePc > (ULONG_PTR)MmHighestUserAddress &&
+            KiPcToFileHeader(FramePc, &LdrEntry, FALSE, &InSystem))
+            {
+                KeBugCheckUnicodeToAnsi(&LdrEntry->BaseDllName,
+                                        AnsiName,
+                                        sizeof(AnsiName));
+                FramePc = (PVOID)((ULONG_PTR)FramePc - (ULONG_PTR)LdrEntry->DllBase);
+                sprintf(Buffer, "%p <%s:%p>\r\n", Frame, AnsiName, FramePc);
```

Comment:
I must have either been really lazy or copied this from somewhere else -- `sprintf` is not allow in new code, please use an ntstrsafe equivalent (e.g. RtlStringCbPrintfA).

Reply:
No problem